sql/core/src/main/scala/org/apache/spark/sql/execution/CacheManager.scala

```diff
@@ -156,10 +165,27 @@ private[sql] class CacheManager(sqlContext: SQLContext) extends Logging {
    * function will over invalidate.
    */
   private[sql] def invalidateCache(plan: LogicalPlan): Unit = writeLock {
-    cachedData.foreach {
-      case data if data.plan.collect { case p if p.sameResult(plan) => p }.nonEmpty =>
-        data.cachedRepresentation.recache()
-      case _ =>
+    var i = 0
+    var locatedIdx = -1
+    // find the index of the cached data, according to the specified logical plan
+    while (i < cachedData.length && locatedIdx < 0) {
+      cachedData(i) match {
+        case data if data.plan.collect { case p if p.sameResult(plan) => p }.nonEmpty =>
+          locatedIdx = i
+        case _ =>
+      }
+      i += 1
+    }
+
+    if (locatedIdx >= 0) {
+      // if the cached data exists, remove it from the cache data list, as we need to
+      // re-generate the spark plan, and we don't want the this to be used during the
+      // re-generation
```

Comment:
In `SQLContext.QueryExecution`, we always try to load the cached `PhysicalRDD` according to the logical plan first, if we don't remove it from the cached list, then we will always get the stale one, which is not what we want, right?